### ISSUE TITLE
SQLite dialect does not understand partial indexes

### ISSUE THREAD

**Issue by joeljuca** [sqlfluff/sqlfluff#4830]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Given the following SQLite table with a partial index:

```sql
-- create-posts.sql

create table posts (
  is_active integer not null default 1,
  title text not null,

  check (is_active in (0, 1)) -- FYI: SQLite does not have a boolean type
);

create index idx0 on posts (title) where is_active = 1;
```

I get the following SQLFluff error:

```
(.venv)  $ sqlfluff lint create-posts.sql
== [create-posts.sql] FAIL
L:  10 | P:  36 |  PRS | Line 10, Position 36: Found unparsable section: 'where
                       | is_active = 1;'
WARNING: Parsing errors found and dialect is set to 'sqlite'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

See: https://www.sqlite.org/partialindex.html

### Expected Behaviour

SQLFluff should understand and parse partial indexes.

### Observed Behaviour

It does not work.

### How to reproduce

Given the following SQLite table with a partial index:

```sql
-- create-posts.sql

create table posts (
  is_active integer not null default 1,
  title text not null,

  check (is_active in (0, 1)) -- FYI: SQLite does not have a boolean type
);

create index idx0 on posts (title) where is_active = 1;
```

I get the following SQLFluff error:

```
(.venv)  $ sqlfluff lint create-posts.sql
== [create-posts.sql] FAIL
L:  10 | P:  36 |  PRS | Line 10, Position 36: Found unparsable section: 'where
                       | is_active = 1;'
WARNING: Parsing errors found and dialect is set to 'sqlite'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

See: https://www.sqlite.org/partialindex.html

### Dialect

SQLite

### Version

2.0.7

### Configuration

```ini
; SQLFluff config file
; See https://docs.sqlfluff.com/en/stable/configuration.html#defaultconfig

[sqlfluff]
encoding = utf-8
dialect = sqlite
templater = raw

[sqlfluff:indentation]
indent_unit = space
tab_space_size = 2
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
